Adapted from https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb 

##### Copyright 2019 The TensorFlow Authors.

In [1]:
#@title Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

In [2]:
#@title MIT License
#
# Copyright (c) 2017 François Chollet
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

# Text classification with TensorFlow Hub: Movie reviews

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/tutorials/keras/text_classification_with_hub"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/docs/blob/master/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
  <td>
    <a href="https://storage.googleapis.com/tensorflow_docs/docs/site/en/tutorials/keras/text_classification_with_hub.ipynb"><img src="https://www.tensorflow.org/images/download_logo_32px.png" />Download notebook</a>
  </td>
</table>


This notebook classifies movie reviews as *positive* or *negative* using the text of the review. This is an example of *binary*—or two-class—classification, an important and widely applicable kind of machine learning problem.

The tutorial demonstrates the basic application of transfer learning with TensorFlow Hub and Keras.

We'll use the [IMDB dataset](https://www.tensorflow.org/api_docs/python/tf/keras/datasets/imdb) that contains the text of 50,000 movie reviews from the [Internet Movie Database](https://www.imdb.com/). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and testing sets are *balanced*, meaning they contain an equal number of positive and negative reviews. 

This notebook uses [tf.keras](https://www.tensorflow.org/guide/keras), a high-level API to build and train models in TensorFlow, and [TensorFlow Hub](https://www.tensorflow.org/hub), a library and platform for transfer learning. For a more advanced text classification tutorial using `tf.keras`, see the [MLCC Text Classification Guide](https://developers.google.com/machine-learning/guides/text-classification/).

### Local Installation
Execute the following from the `\Scripts` directory of the Python installation.

```
pip install numpy
pip install tensorflow
pip install "C:\path\googleapis_common_protos-1.51.0-py3-none-any.whl"
pip install typing
pip install tensorflow-datasets
pip install ipywidgets
jupyter nbextension enable --py widgetsnbextension --sys-prefix
```

Restart jupyter if it is running


In [3]:
from __future__ import absolute_import, division, print_function, unicode_literals

import numpy as np

try:
  # %tensorflow_version only exists in Colab.
  %tensorflow_version 2.x
except Exception:
  pass
import tensorflow as tf

#!pip install tensorflow-hub
#!pip install tfds-nightly
import tensorflow_hub as hub
import tensorflow_datasets as tfds

print("Version: ", tf.__version__)
print("Eager mode: ", tf.executing_eagerly())
print("Hub version: ", hub.__version__)
print("GPU is", "available" if tf.config.experimental.list_physical_devices("GPU") else "NOT AVAILABLE")

Version:  2.1.0
Eager mode:  True
Hub version:  0.7.0
GPU is NOT AVAILABLE


## Download the IMDB dataset

The IMDB dataset is available on [imdb reviews](https://www.tensorflow.org/datasets/catalog/imdb_reviews) or on [TensorFlow datasets](https://www.tensorflow.org/datasets). The following code downloads the IMDB dataset to your machine (or the colab runtime):

The dataset is downloaded to `%userprofile%\tensorflow_datasets\imdb_reviews`. 

In [5]:
# Split the training set into 60% and 40%, so we'll end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews", 
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

In [6]:
train_data_list = [x for x in train_data]
print("Number of reviews in train_data", len(train_data_list))

Number of reviews in train_data 15000


## Explore the data 

Let's take a moment to understand the format of the data. Each example is a sentence representing the movie review and a corresponding label. The sentence is not preprocessed in any way. The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

Let's print first 10 examples.

I replaced the original code withcode that prints a short version of the first three reviews.

In [7]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(10)))
train_examples_batch
printable_examples = [x.numpy()[:150] for x in train_examples_batch[:3]]
for x in printable_examples:
    print(x)

b"This is a big step down after the surprisingly enjoyable original. This sequel isn't nearly as fun as part one, and it instead spends too much time on"
b"Perhaps because I was so young, innocent and BRAINWASHED when I saw it, this movie was the cause of many sleepless nights for me. I haven't seen it si"
b'Hood of the Living Dead had a lot to live up to even before the opening credits began. First, any play on "...of the living dead" invokes His Holiness'


Let's also print the first 10 labels.

In [8]:
train_labels_batch

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([0, 0, 1, 0, 1, 0, 1, 1, 1, 0], dtype=int64)>

## Build the model

The neural network is created by stacking layers—this requires three main architectural decisions:

* How to represent the text?
* How many layers to use in the model?
* How many *hidden units* to use for each layer?

In this example, the input data consists of sentences. The labels to predict are either 0 or 1.

One way to represent the text is to convert sentences into embeddings vectors. We can use a pre-trained text embedding as the first layer, which will have three advantages:

*   we don't have to worry about text preprocessing,
*   we can benefit from transfer learning,
*   the embedding has a fixed size, so it's simpler to process.

For this example we will use a **pre-trained text embedding model** from [TensorFlow Hub](https://www.tensorflow.org/hub) called [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1).

There are three other pre-trained models to test for the sake of this tutorial:

* [google/tf2-preview/gnews-swivel-20dim-with-oov/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim-with-oov/1) - same as [google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1), but with 2.5% vocabulary converted to OOV buckets. This can help if vocabulary of the task and vocabulary of the model don't fully overlap.
* [google/tf2-preview/nnlm-en-dim50/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim50/1) - A much larger model with ~1M vocabulary size and 50 dimensions.
* [google/tf2-preview/nnlm-en-dim128/1](https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1) - Even larger model with ~1M vocabulary size and 128 dimensions.

Let's first create a Keras layer that uses a TensorFlow Hub model to embed the sentences, and try it out on a couple of input examples. Note that no matter the length of the input text, the output shape of the embeddings is: `(num_examples, embedding_dimension)`.

In [9]:
#embedding = "https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1"
embedding = "https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1"
hub_layer = hub.KerasLayer(embedding, input_shape=[], 
                           dtype=tf.string, trainable=True)
hub_layer(train_examples_batch[:3])

<tf.Tensor: shape=(3, 128), dtype=float32, numpy=
array([[ 1.1329565 , -0.03431067,  0.12162483,  0.23909318, -0.00590096,
         0.01708133,  0.08015746, -0.24174148,  0.02650847, -0.04987909,
         0.11247347, -0.5601333 , -0.19981343, -0.33171475, -0.21482827,
        -0.0408943 , -0.07008895,  0.04761497, -0.4271519 ,  0.5916435 ,
         0.20662342, -0.01224876,  0.02093347,  0.08682311,  0.1036322 ,
        -0.27562743,  0.08033531, -0.1508889 , -0.21847422,  0.06831451,
        -0.14723632, -0.01117621,  0.25246456,  0.10763772,  0.16948967,
        -0.0666896 , -0.2209269 , -0.3792632 , -0.15618715,  0.20775793,
        -0.24780062,  0.05174338,  0.04574278, -0.03784208,  0.09963844,
         0.2984763 ,  0.04106063, -0.1306813 , -0.23743384,  0.07203843,
         0.09253472, -0.13241911, -0.29098544,  0.1412372 , -0.20057504,
        -0.02892039, -0.14496696, -0.11467611,  0.35817885,  0.137189  ,
        -0.0982758 ,  0.04868827,  0.01760143, -0.08307212, -0.0034984 ,
 

### Explore Word Embeddings

The https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1 `SavedModel 2.0` has a vocabulary of 20,000 tokens and one out-of-vocabulary bucket for unknown tokens. The result is a tensor with 20,001 rows and 20 columns. The https://tfhub.dev/google/tf2-preview/nnlm-en-dim128/1 `SavedModel 2.0` has a vocabulary of 973,771 tokens and 128 dimensions.

In [10]:
hub_layer_variables = hub_layer.variables
print(hub_layer_variables[0].shape)

(973771, 128)


Let's now build the full model:

In [11]:
model = tf.keras.Sequential()
model.add(hub_layer)
model.add(tf.keras.layers.Dense(16, activation='relu'))
model.add(tf.keras.layers.Dense(1, activation='sigmoid'))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
keras_layer (KerasLayer)     (None, 128)               124642688 
_________________________________________________________________
dense (Dense)                (None, 16)                2064      
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 17        
Total params: 124,644,769
Trainable params: 124,644,769
Non-trainable params: 0
_________________________________________________________________


The layers are stacked sequentially to build the classifier:

1. The first layer is a TensorFlow Hub layer. This layer uses a pre-trained Saved Model to map a sentence into its embedding vector. The pre-trained text embedding model that we are using ([google/tf2-preview/gnews-swivel-20dim/1](https://tfhub.dev/google/tf2-preview/gnews-swivel-20dim/1)) splits the sentence into tokens, embeds each token and then combines the embedding. The resulting dimensions are: `(num_examples, embedding_dimension)`.
2. This fixed-length output vector is piped through a fully-connected (`Dense`) layer with 16 hidden units.
3. The last layer is densely connected with a single output node. Using the `sigmoid` activation function, this value is a float between 0 and 1, representing a probability, or confidence level.

Let's compile the model.

### Loss function and optimizer

A model needs a loss function and an optimizer for training. Since this is a binary classification problem and the model outputs a probability (a single-unit layer with a sigmoid activation), we'll use the `binary_crossentropy` loss function. 

This isn't the only choice for a loss function, you could, for instance, choose `mean_squared_error`. But, generally, `binary_crossentropy` is better for dealing with probabilities—it measures the "distance" between probability distributions, or in our case, between the ground-truth distribution and the predictions.

Later, when we are exploring regression problems (say, to predict the price of a house), we will see how to use another loss function called mean squared error.

Now, configure the model to use an optimizer and a loss function:

In [12]:
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

## Train the model

Train the model for 20 epochs in mini-batches of 512 samples. This is 20 iterations over all samples in the `x_train` and `y_train` tensors. While training, monitor the model's loss and accuracy on the 10,000 samples from the validation set:

In [16]:
history = model.fit(train_data.shuffle(10000).batch(1512),
                    epochs=200,
                    validation_data=validation_data.batch(1512),
                    verbose=1, use_multiprocessing=True, workers=3)

Epoch 1/200
10/10 [==============================] - 22s 2s/step - loss: 0.0382 - accuracy: 0.9968 - val_loss: 0.3245 - val_accuracy: 0.8730
Epoch 2/200
10/10 [==============================] - 22s 2s/step - loss: 0.0328 - accuracy: 0.9976 - val_loss: 0.3302 - val_accuracy: 0.8719
Epoch 3/200
10/10 [==============================] - 21s 2s/step - loss: 0.0284 - accuracy: 0.9983 - val_loss: 0.3365 - val_accuracy: 0.8714
Epoch 4/200
10/10 [==============================] - 20s 2s/step - loss: 0.0247 - accuracy: 0.9989 - val_loss: 0.3428 - val_accuracy: 0.8714
Epoch 5/200
10/10 [==============================] - 22s 2s/step - loss: 0.0216 - accuracy: 0.9991 - val_loss: 0.3488 - val_accuracy: 0.8711
Epoch 6/200
10/10 [==============================] - 21s 2s/step - loss: 0.0189 - accuracy: 0.9994 - val_loss: 0.3545 - val_accuracy: 0.8704
Epoch 7/200
10/10 [==============================] - 20s 2s/step - loss: 0.0166 - accuracy: 0.9995 - val_loss: 0.3599 - val_accuracy: 0.8696
Epoch 8/200
1

10/10 [==============================] - 20s 2s/step - loss: 8.4268e-04 - accuracy: 1.0000 - val_loss: 0.5243 - val_accuracy: 0.8619
Epoch 59/200
10/10 [==============================] - 20s 2s/step - loss: 8.1811e-04 - accuracy: 1.0000 - val_loss: 0.5264 - val_accuracy: 0.8623
Epoch 60/200
10/10 [==============================] - 21s 2s/step - loss: 7.9460e-04 - accuracy: 1.0000 - val_loss: 0.5280 - val_accuracy: 0.8619
Epoch 61/200
10/10 [==============================] - 20s 2s/step - loss: 7.7100e-04 - accuracy: 1.0000 - val_loss: 0.5300 - val_accuracy: 0.8619
Epoch 62/200
10/10 [==============================] - 21s 2s/step - loss: 7.4899e-04 - accuracy: 1.0000 - val_loss: 0.5318 - val_accuracy: 0.8622
Epoch 63/200
10/10 [==============================] - 21s 2s/step - loss: 7.2813e-04 - accuracy: 1.0000 - val_loss: 0.5335 - val_accuracy: 0.8618
Epoch 64/200
10/10 [==============================] - 20s 2s/step - loss: 7.0822e-04 - accuracy: 1.0000 - val_loss: 0.5353 - val_accuracy

10/10 [==============================] - 24s 2s/step - loss: 2.4162e-04 - accuracy: 1.0000 - val_loss: 0.6037 - val_accuracy: 0.8612
Epoch 115/200
10/10 [==============================] - 23s 2s/step - loss: 2.3766e-04 - accuracy: 1.0000 - val_loss: 0.6047 - val_accuracy: 0.8611
Epoch 116/200
10/10 [==============================] - 23s 2s/step - loss: 2.3350e-04 - accuracy: 1.0000 - val_loss: 0.6058 - val_accuracy: 0.8611
Epoch 117/200
10/10 [==============================] - 23s 2s/step - loss: 2.2983e-04 - accuracy: 1.0000 - val_loss: 0.6069 - val_accuracy: 0.8611
Epoch 118/200
10/10 [==============================] - 22s 2s/step - loss: 2.2589e-04 - accuracy: 1.0000 - val_loss: 0.6080 - val_accuracy: 0.8611
Epoch 119/200
10/10 [==============================] - 24s 2s/step - loss: 2.2232e-04 - accuracy: 1.0000 - val_loss: 0.6091 - val_accuracy: 0.8611
Epoch 120/200
10/10 [==============================] - 22s 2s/step - loss: 2.1867e-04 - accuracy: 1.0000 - val_loss: 0.6101 - val_ac

Epoch 170/200
10/10 [==============================] - 22s 2s/step - loss: 1.0790e-04 - accuracy: 1.0000 - val_loss: 0.6565 - val_accuracy: 0.8606
Epoch 171/200
10/10 [==============================] - 21s 2s/step - loss: 1.0666e-04 - accuracy: 1.0000 - val_loss: 0.6573 - val_accuracy: 0.8606
Epoch 172/200
10/10 [==============================] - 22s 2s/step - loss: 1.0531e-04 - accuracy: 1.0000 - val_loss: 0.6580 - val_accuracy: 0.8605
Epoch 173/200
10/10 [==============================] - 22s 2s/step - loss: 1.0403e-04 - accuracy: 1.0000 - val_loss: 0.6588 - val_accuracy: 0.8605
Epoch 174/200
10/10 [==============================] - 21s 2s/step - loss: 1.0278e-04 - accuracy: 1.0000 - val_loss: 0.6597 - val_accuracy: 0.8606
Epoch 175/200
10/10 [==============================] - 22s 2s/step - loss: 1.0151e-04 - accuracy: 1.0000 - val_loss: 0.6605 - val_accuracy: 0.8604
Epoch 176/200
10/10 [==============================] - 23s 2s/step - loss: 1.0035e-04 - accuracy: 1.0000 - val_loss: 0

## Evaluate the model

And let's see how the model performs. Two values will be returned. Loss (a number which represents our error, lower values are better), and accuracy.

In [17]:
results = model.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model.metrics_names, results):
  print("%s: %.3f" % (name, value))

loss: 0.742
accuracy: 0.846


This fairly naive approach achieves an accuracy of about 87%. With more advanced approaches, the model should get closer to 95%.

## Further reading

For a more general way to work with string inputs and for a more detailed analysis of the progress of accuracy and loss during training, take a look [here](https://www.tensorflow.org/tutorials/keras/basic_text_classification).